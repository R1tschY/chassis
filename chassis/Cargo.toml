[package]
name = "chassis"
version = "0.1.0"
authors = ["R1tschY <r1tschy@posteo.de>"]
edition = "2018"
license = "MIT OR Apache-2.0"
description = "Compile time dependency injection framework"
repository = "https://github.com/R1tschY/chassis"
readme = "README.md"
keywords = ["dependency-injection"]

[lib]
proc-macro = true

[dependencies]
syn = { version = "1.0", features = ["full"] }
quote = "1.0"
proc-macro2 = "1.0"

[dev-dependencies]
assert_matches = "1.3"

[build-dependencies]
version_check = "0.9"