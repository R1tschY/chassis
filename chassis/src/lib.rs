#![cfg_attr(nightly_diagnostics, feature(proc_macro_diagnostic, proc_macro_span))]

#[macro_use]
extern crate quote;
#[macro_use]
extern crate syn;

use proc_macro::TokenStream;

use syn::export::TokenStream2;

use crate::codegen::codegen_component_impl;
use crate::container::IocContainer;
use crate::errors::{codegen_errors, ChassisError, ChassisResult};
use crate::parse::parse_block;
use syn::spanned::Spanned;

mod codegen;
mod container;
mod diagnostic;
mod errors;
mod model;
mod parse;
mod syn_ext;
mod utils;

#[proc_macro_attribute]
pub fn integration(_args: TokenStream, input: TokenStream) -> TokenStream {
    let mod_block: syn::ItemMod = parse_macro_input!(input);

    match parse_integration(_args, mod_block) {
        Ok(tokens) => tokens.into(),
        Err(err) => codegen_errors(err).into(),
    }
}

fn parse_integration(
    _args: TokenStream,
    mut mod_block: syn::ItemMod,
) -> ChassisResult<TokenStream2> {
    let mut mod_impl = match &mut mod_block.content {
        Some((_, items)) => items,
        None => {
            return Err(ChassisError::IllegalInput(
                "Expected module implementation when using integration attribute".to_string(),
                mod_block.span().clone(),
            ))
        }
    };

    // Parse components and modules
    let block = parse_block(&mut mod_impl)?;

    // analyse
    let modules = block.modules;
    let mut container = IocContainer::new();
    for module in modules {
        container.add_module(module)?;
    }

    // generate
    let component_impls = block
        .components
        .into_iter()
        .map(|comp| codegen_component_impl(comp, &container))
        .collect::<ChassisResult<Vec<TokenStream2>>>()?;

    // generate result
    let mod_name = &mod_block.ident;
    let mod_vis = &mod_block.vis;
    Ok(quote! {
        #mod_vis mod #mod_name {
            #(#mod_impl)*

            #(#component_impls)*
        }
    })
}
