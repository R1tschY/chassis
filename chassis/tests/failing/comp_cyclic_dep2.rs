use chassis::integration;

pub struct Dummy1;
pub struct Dummy2;

#[integration]
mod mut_self {
    use super::*;

    pub struct DummyProvider;

    impl DummyProvider {
        pub fn provide_dummy2(dummy: Dummy1) -> Dummy2 {
            Dummy2
        }

        pub fn provide_dummy1(dummy: Dummy2) -> Dummy1 {
            Dummy1
        }
    }

    pub trait DummyFactory {
        fn resolve_dummy(&self) -> Dummy1;
    }
}

fn main() {}
