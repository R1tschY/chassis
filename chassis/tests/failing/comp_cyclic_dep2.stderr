error: Cyclic dependency for `Dummy1`
  --> $DIR/comp_cyclic_dep2.rs:13:38
   |
13 |         pub fn provide_dummy2(dummy: Dummy1) -> Dummy2 {
   |                                      ^^^^^^
   |
help: required for `Dummy2`
  --> $DIR/comp_cyclic_dep2.rs:17:38
   |
17 |         pub fn provide_dummy1(dummy: Dummy2) -> Dummy1 {
   |                                      ^^^^^^
help: required for `Dummy1`
  --> $DIR/comp_cyclic_dep2.rs:23:36
   |
23 |         fn resolve_dummy(&self) -> Dummy1;
   |                                    ^^^^^^
